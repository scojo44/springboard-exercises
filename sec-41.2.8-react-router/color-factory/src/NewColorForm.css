.NewColorForm {
  margin: 1rem;
}
.NewColorForm input {
  color: white;
  background-color: #444;
  border: 3px inset #444;
}
.NewColorForm button {
  color: white;
  background-color: #444;
  font-weight: bold;
  padding: .3rem .5rem;
  border: 3px outset #444;
  border-radius: .5rem;
}