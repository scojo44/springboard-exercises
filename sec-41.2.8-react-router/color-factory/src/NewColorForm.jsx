import React, {useState} from "react"
import {v4 as uuid} from 'uuid'
import { useNavigate } from "react-router-dom"
import './NewColorForm.css'

const EMPTY_FORM = {
  name: '',
  value: ''
};

const NewColorForm = ({addColor}) => {
  const [color, setColor] = useState(EMPTY_FORM);
  const navigate = useNavigate();

  return (
    <form onSubmit={handleSubmit} className="NewColorForm">
      <p>
        <label htmlFor="name">Color Name: </label>
        <input type="text" id="name" name="name" onChange={handleChange} value={color.name} />
      </p>
      <p>
        <label htmlFor="value">Color Value: </label>
        <input type="color" id="value" name="value" onChange={handleChange} value={color.value} />
      </p>
      <p>
        <button type="submit">Add Color</button>
      </p>
    </form>
  );

  function handleSubmit(e) {
    addColor({id: uuid(), ...color});
    setColor(() => EMPTY_FORM);
    navigate('/colors');
  }

  function handleChange(e) {
    const {name, value} = e.target;
    setColor(color => ({
      ...color,
      [name]: value
    }));
  }
};

export default NewColorForm
