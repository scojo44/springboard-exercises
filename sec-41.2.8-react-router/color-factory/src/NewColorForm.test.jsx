import {render, fireEvent} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import NewColorForm from './NewColorForm'

describe('NewColorForm Tests', () => {
  const mockAddColor = vi.fn();

  it('Renders without crashing', () => {
    render(
      <MemoryRouter>
        <NewColorForm addColor={mockAddColor} />
      </MemoryRouter>
    );
  });

  it('Matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <NewColorForm addColor={mockAddColor} />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Calls the passed in function on submit', () => {
    const {getByText, getByLabelText} = render(
      <MemoryRouter>
        <NewColorForm addColor={mockAddColor} />
      </MemoryRouter>
    );
    const colorName = getByLabelText('Color Name:');
    const colorValue = getByLabelText('Color Value:');
    const button = getByText('Add Color');

    fireEvent.change(colorName, {target: {value: 'mediumpurple'}});
    fireEvent.change(colorValue, {target: {value: '#9370db'}});
    fireEvent.click(button);

    expect(mockAddColor).toHaveBeenCalledWith({
      id: expect.stringMatching(/[a-z0-9-]+/), // UUID
      name: 'mediumpurple',
      value: '#9370db'
    })
  });
});
