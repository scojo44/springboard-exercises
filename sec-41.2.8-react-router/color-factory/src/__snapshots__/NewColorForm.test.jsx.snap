// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewColorForm Tests > Matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewColorForm"
  >
    <p>
      <label
        for="name"
      >
        Color Name: 
      </label>
      <input
        id="name"
        name="name"
        type="text"
        value=""
      />
    </p>
    <p>
      <label
        for="value"
      >
        Color Value: 
      </label>
      <input
        id="value"
        name="value"
        type="color"
        value=""
      />
    </p>
    <p>
      <button
        type="submit"
      >
        Add Color
      </button>
    </p>
  </form>
</DocumentFragment>
`;
